sdb/commands/member.py

```diff
@@ -31,16 +32,145 @@ class Member(sdb.Command):
     def _init_argparse(self, parser: argparse.ArgumentParser) -> None:
         parser.add_argument("members", nargs="+", metavar="<member>")
 
-    def call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
-        for obj in objs:
-            for member in self.args.members:
+    @staticmethod
+    def _lex_member_tokens(member_expr: str) -> List[str]:
+        return list(filter(None, re.split(r'(\.|->|\[|\])', member_expr)))
+
+    def _parse_member_tokens(self, tokens: List[str]) -> List[Tuple[str, str]]:
+        terms = []
+        while tokens:
+
+            is_index = False
+            identifier = ""
+            idx = ""
+            if tokens[0] == "." or tokens[0] == "->":
+                sep = tokens[0]
+                identifier = tokens[1]
+                tokens = tokens[2:]
+            elif tokens[0] == '[':
+                if len(tokens) < 3 or tokens[2] != ']':
+                    raise sdb.CommandError(self.name,
+                                           "incomplete array expression")
+                if not tokens[1].isdigit():
+                    raise sdb.CommandError(
+                        self.name,
+                        "incorrect index: '{}' is not a number".format(
+                            tokens[1]))
+                is_index = True
+                idx = tokens[1]
+                sep = "[]"
+                tokens = tokens[3:]
+            else:
+                sep = ""
+                identifier = tokens[0]
+                tokens = tokens[1:]
+
+            if not is_index and not identifier.isidentifier():
+                raise sdb.CommandError(
+                    self.name,
+                    "{} is not an acceptable identifier".format(identifier))
+
+            if not is_index:
+                assert idx == ""
+                terms.append((sep, identifier))
+            else:
+                assert identifier == ""
+                terms.append((sep, idx))
+        return terms
+
+    @staticmethod
+    def _typedef_to_base_kind(type_: drgn.Type) -> drgn.TypeKind:
+        while type_.kind == drgn.TypeKind.TYPEDEF:
+            type_ = type_.type
+        return type_.kind
+
+    def _validate_type_dereference(self, obj: drgn.Object, sep: str) -> None:
+        kind = self._typedef_to_base_kind(obj.type_)
+
+        # This is the first term, no need to do validation.
+        if not sep:
+            return
+
+        if kind == drgn.TypeKind.STRUCT and sep != ".":
+            raise sdb.CommandError(
+                self.name,
+                "'{}' is a struct - use the dot(.) notation for member access".
+                format(obj.type_))
+
+        if kind == drgn.TypeKind.POINTER and sep != "->" and sep != '.':
```

Comment:
How do I access the entries in this array:
```
> spa | member spa_zio_taskq[0][0].stqs_taskq
*(taskq_t **)0xffff931b9817aad8 = 0xffff931b98a1ac00
> spa | member spa_zio_taskq[0][0].stqs_taskq[0]
sdb: member: 'taskq_t **' is a pointer - use the arrow(->) notation for member access
```
? and/or should I be using `member` to access these entries, or something else?
 
Similarly, if the variable is a `char *`, it might be useful to use array notation to get at the contents of the string; e.g.
```
> spa | member spa_comment[0]
sdb: member: 'char *' is a pointer - use the arrow(->) notation for member access
```

Reply:
After giving it some though I think allowing the array notation for pointers is not a bad idea and we really cannot do anything else - C's type system and memory mode is the way it is.